app/controllers/logins_controller.rb

```diff
@@ -101,47 +100,43 @@ def totp
   end
 
   def complete
+    # Clear the flash - this prevents the error message showing up after an unsuccessful -> successful login
+    flash.clear
+
+    service = ProcessLoginService.new(login: @login)
 
     case params[:method]
     when "webauthn"
-      webauthn_credential = WebAuthn::Credential.from_get(JSON.parse(params[:credential]))
-      stored_credential = @user.webauthn_credentials.find_by!(webauthn_id: webauthn_credential.id)
-
-      webauthn_credential.verify(
-        session[:webauthn_challenge],
-        public_key: stored_credential.public_key,
-        sign_count: stored_credential.sign_count
+      ok = service.process_webauthn(
+        raw_credential: params[:credential],
+        challenge: session[:webauthn_challenge]
       )
 
-      stored_credential.update!(sign_count: webauthn_credential.sign_count)
-
-      session[:login_preference] = "webauthn" if params[:remember] == "true"
-
-      @login.update(authenticated_with_webauthn: true)
-
+      unless ok
+        redirect_to(auth_users_path, flash: { error: service.errors.full_messages.to_sentence })
+        return
+      end
     when "login_code"
-      UserService::ExchangeLoginCodeForUser.new(
-        user_id: @login.user.id,
-        login_code: params[:login_code],
-        sms: params[:sms],
-      ).run
-
-      if params[:sms]
-        @login.update(authenticated_with_sms: true)
-      else
-        @login.update(authenticated_with_email: true)
+      ok = service.process_login_code(
+        code: params[:login_code],
+        sms: ActiveRecord::Type::Boolean.new.cast(params[:sms])
```

Comment:
```suggestion
        sms: !!params[:sms]
```
This is mainly a personal preference

Reply:
Those aren't strictly equivalent though (e.g. `""` is `true` if you use `!!` but `false` if you cast)